File: Trees/BinaryTreesC++/verticalOrderTraversal.cpp
#include <bits/stdc++.h>
using namespace std;


class node{
	public:
		int data;
		node* left;
		node* right;
		node(int d){
			this->data = d;
			this->left = NULL;
			this->right = NULL;
		}
};
node* buildTree(node* root){	
	int data; 
	cin>>data ;
	root = new node(data);
	if(data == -1){
		return NULL;
	}
	

	root->left = buildTree(root->left);
	root->right = buildTree(root->right);
	return root;

}

vector<int> verticalTraversal(node* root){
	map<int , map<int,vector<int>>> nodes ;
	vector<int> ans ;

	if(root == NULL){
		return ans;
	}

	queue<pair<node*,pair<int , int >>> q ;

	q.push(make_pair(root , make_pair( 0 , 0 )));

	while(!q.empty()){
		pair<node*,pair<int , int >> temp = q.front();
		q.pop();
		node* front  = temp.first;
		int hd = temp.second.first;
		int level =  temp.second.second ;
		int data = front->data;
		nodes[hd][level].push_back(data);

		if(front->left != NULL){
			q.push(make_pair(front->left,make_pair(hd-1,level+1)));
		}
		if(front->right != NULL){
			q.push(make_pair(front->right,make_pair(hd+1,level+1)));
		}
	}
	for(auto i : nodes){
		for(auto j : i.second){
			for(auto k : j.second){
				ans.push_back(k);
			}
		}
	}
	return ans ;
}

int main(){
	node* root = NULL ;
	root = buildTree(root);
	vector<int> ans  = verticalTraversal(root);
	for(auto i : ans){
		cout<< i <<" ";
	}
}

File: Trees/BinaryTreeJava/src/binarytree/VerticalOrderTraversal.java
package binarytree;

import java.util.*;

public class VerticalOrderTraversal {
    public static void main(String[] args) {
        Node root = null;
        DiameterOrWidth.Height h = new DiameterOrWidth.Height();
        BinaryTree s = new BinaryTree();
        root = s.buildTree(root);
        ArrayList<Integer> ans  = verticalOrder(root);
        System.out.println(ans);
    }
    
    static class NodeHD{
        Node node ;
        int hd ;
        
        NodeHD(Node node , int hd){
            this.node = node ;
            this.hd = hd ;
        }
        NodeHD(){
            this.node = null ;
            this.hd = 0 ;
        }
    }
    
    public static ArrayList<Integer> verticalOrder(Node root) {
        Map<Integer,ArrayList<Integer>> nodes =  new TreeMap<>();
        ArrayList<Integer> ans =  new ArrayList<>();
        if(root == null){
            return ans ;
        }
        Queue<NodeHD> q = new LinkedList<>();
        NodeHD first = new NodeHD(root,0);
        q.add(first);

        int min = 0 ;
        int max = 0;
        
        while(!q.isEmpty()){
            NodeHD temp = q.remove() ;
            int hd = temp.hd;
            Node fNode = temp.node;
            nodes.putIfAbsent(hd,new ArrayList<>());
            nodes.get(hd).add(fNode.data);
            if(fNode.left != null){
                q.add(new NodeHD(fNode.left , hd-1));
                min = min - 1;
            }
            if(fNode.right != null){
                q.add(new NodeHD(fNode.right , hd+1));
                max =  max + 1;
            }
        }
//        for (int i = min+1; i < max ;i++){
//            ans.addAll(nodes.get(i));
//        }

        for (Map.Entry<Integer, ArrayList<Integer>> i : nodes.entrySet()) {
            ArrayList<Integer> a = i.getValue();
            ans.addAll(a);
        }

        return ans ;
    }
}
